File: miscellaneous/swap_nodes.py
'''
leetcode problem 24 - swap nodes in pairs
'''
class ListNode:
	def __init__(self, val = 0, next = None):
		self.val = val
		self.next = next


class Solution:
	def swapPairs(self, head):
		if head is None or head.next is None:
			return head
		else:
			temp = head
			head = head.next
			temp.next = head.next
			head.next = temp

			temp.next = self.swapPairs(temp.next)
	
		return head

File: miscellaneous/swap_nodes.cpp
//leetcode problem 24 - swapping nodes in pairs

struct ListNode{
	int value;
	ListNode *next;
	ListNode(): value(0), next(nullptr){}
	ListNode(int x):value(x), next(nullptr){}
	ListNode(int x, ListNode *next):value(x), next(next){}
};

ListNode* swapPairs(ListNode* head){
	if(head == NULL || head -> next == NULL)
		return head;
	ListNode *node = head;
	ListNode *curr = head;

	node = curr -> next;
	curr -> next = swapPairs(head -> next -> next);
	node -> next = curr;
	head = node;
	return head;
}
